vespalib/src/vespa/vespalib/net/tls/transport_security_options_reading.cpp

```diff
@@ -0,0 +1,102 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#include "transport_security_options_reading.h"
+#include <vespa/vespalib/data/slime/slime.h>
+#include <vespa/vespalib/util/exceptions.h>
+#include <vespa/vespalib/io/fileutil.h>
+#include <vespa/vespalib/io/mapped_file_input.h>
+#include <vespa/vespalib/data/memory_input.h>
+
+namespace vespalib::net::tls {
+
+/*
+
+ Proposed JSON format for TLS configuration file:
+
+{
+  "files": {
+    "private-key": "myhost.key",
+    "ca-certificates": "my_cas.pem",
+    "certificates": "certs.pem"
+  },
+  // for later:
+  "peer-taggers": [
+    {
+      "requirements":[
+        {
+          "field": "SAN"
+          "must-match": "DNS:foo.bar.baz.*"
+        }
+      ],
+      "tags": ["cluster-peers", "config-server"] // or "roles"? Avoid ambiguities with Athenz concepts
+    },
+    {
+      "requirements":[
+        { "field":"CN", "must-match": "config.blarg.*"}
+      ],
+      "tags": ["config-server"]
+    }
+  ]
+}
+
+ */
+
+using namespace slime;
+
+namespace {
+
+constexpr const char* files_field = "files";
+constexpr const char* private_key_field = "private-key";
+constexpr const char* ca_certs_field = "ca-certificates";
+constexpr const char* certs_field = "certificates";
+
+void verify_referenced_file_exists(const vespalib::string& file_path) {
+    if (!fileExists(file_path)) {
+        throw IllegalArgumentException(make_string("File '%s' referenced by TLS config does not exist", file_path.c_str()));
+    }
+}
+
+vespalib::string load_file_referenced_by_field(const Cursor& cursor, const char* field) {
+    auto file_path = cursor[field].asString().make_string();
+    if (file_path.empty()) {
+        throw IllegalArgumentException(make_string("TLS config field '%s' has not been set", field));
+    }
+    verify_referenced_file_exists(file_path);
+    return File::readAll(file_path);
+}
+
+std::unique_ptr<TransportSecurityOptions> load_from_input(Input& input) {
+    Slime root;
+    auto parsed = JsonFormat::decode(input, root);
+    if (parsed == 0) {
+        throw IllegalArgumentException("Provided TLS config file is not valid JSON");
+    }
+    auto& files = root[files_field];
+    if (files.children() == 0) {
```

Comment:
use fields (object) instead of children (array/object) to be more specific

Reply:
or just use transitive invalid nix behavior to delay error handling to individual files...

Reply:
Moved to `fields()`